### PR TITLE
feat: passkey strategy with resident keys

### PR DESCRIPTION
**Pull request by hperl**
# TODO

* [x] Two distinct strategies, how to prevent passkey + webauthn passkey with too high AAL
	* Passkey is currently always AAL1 and not a second factor.
	* We could disallow using the same authenticator ID (AAGUID) for both webauthn and passkey. However, this could hinder migration from webauthn+passwordless to passkey.
* [x] PassKeys what happens to non-resident keys? For example when migrating WebAuthn passwordless credentials to passkey?
  * This work does not yet consider converting passwordless credentials to a passkey. I think the best way would be to do this with user interaction and insert a "nag screen" after login to prompt the user to create a passkey (and optionally delete/deactivate the webauth passwordless credential). The same technique could be used to onboard users to passkeys regardless of their primary login strategy.
* [x] 2x PassKeys if submission fails
  * In some cases it can be that the passkey is created in the browser but not stored in the database. Ultimately, there is nothing that we can do to completely prevent this, but is should be uncommon, because we store and validate the traits first before we create the passkey.
* [x] WebAuthn.js - use onload for autocomplete
  * Done
* [x] How will the deprecation work of the webauthn strategy?
  * This will need further design and code.
* [x] Clean up `webauthnIdentifierNode`

## Related issue(s)

Docs PR: https://github.com/ory/docs/pull/1626

## Checklist

- [x] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md).
- [x] I have referenced an issue containing the design document if my change
      introduces a new feature.
- [x] I am following the
      [contributing code guidelines](../blob/master/CONTRIBUTING.md#contributing-code).
- [x] I have read the [security policy](../security/policy).
- [x] I confirm that this pull request does not address a security
      vulnerability. If this pull request addresses a security vulnerability, I
      confirm that I got the approval (please contact
      [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push
      the changes.
- [x] I have added tests that prove my fix is effective or that my feature
      works.
- [x] I have added or changed [the documentation](https://github.com/ory/docs).

### CODE REVIEW

File: embedx/config.schema.json
Comment (jonas-jonas) [inline embedx/config.schema.json:822]: @aeneas passkey or passkeys? your call.

Comment (jonas-jonas) [inline embedx/config.schema.json:822]: @aeneasr passkey or passkeys? your call.

Comment (aeneasr) [reply embedx/config.schema.json:822]: PassKeys is the official name so I would stick to that :)

File: selfservice/strategy/passkey/passkey_registration.go
Comment (aeneasr) [inline selfservice/strategy/passkey/passkey_registration.go:182]: I think we should  return an easy to understand error if `webAuthnSess.UserID` is empty. If it is empty, the current validation will pass it as valid

```
	c := i.GetCredentialsOr(identity.CredentialsTypePasskey, &identity.Credentials{})
	if len(c.Identifiers) == 0 {
		return schema.NewMissingIdentifierError()
	}
```

and will lead to an empty string identifier. We had this problem once before in the WebAuthn code.

File: selfservice/strategy/passkey/passkey_login.go
Comment (aeneasr) [inline selfservice/strategy/passkey/passkey_login.go:43]: Shouldn't this only populate for AAL1?

File: selfservice/strategy/passkey/passkey_login.go
Comment (aeneasr) [inline selfservice/strategy/passkey/passkey_login.go:234]: I don't think this is needed here. We're looking for the identity based on the user handle which is a randomly generated string. If this was an email, this could make sense here.

File: selfservice/strategy/passkey/passkey_login.go
Comment (aeneasr) [inline selfservice/strategy/passkey/passkey_login.go:239]: This isn't the correct error? We have a DB error above

**Review by aeneasr — COMMENTED** [review]
Preliminary review - also pushing some code changes now. Can you please add e2e tests for the various flows?

File: ui/node/attributes.go
Comment (aeneasr) [inline ui/node/attributes.go:102]: Carries state that is used in another context. Leaky?

File: ui/node/attributes_input.go
Comment (aeneasr) [inline ui/node/attributes_input.go:152]: Question this

File: ui/node/identifiers.go
Comment (aeneasr) [inline ui/node/identifiers.go:32]: What is `settingsRegister`'s responsibility?

File: ui/node/identifiers.go
Comment (aeneasr) [inline ui/node/identifiers.go:33]: All other names are prefixed with the credential type - please keep consistency?

File: ui/node/identifiers.go
Comment (aeneasr) [inline ui/node/identifiers.go:36]: All other names are prefixed with the credential type - please keep consistency?

File: x/webauthnx/aaguid/aaguid.go
Comment (aeneasr) [inline x/webauthnx/aaguid/aaguid.go:32]: One test maybe? :)

File: x/webauthnx/js/webauthn.js
Comment (aeneasr) [inline x/webauthnx/js/webauthn.js:135]: What happens here? Is the flow simply aborted? Does the user receive some indication why it did not work?

**Review by aeneasr — CHANGES_REQUESTED** [review]

File: identity/credentials_webauthn.go
Comment (aeneasr) [reply identity/credentials_webauthn.go:38]: I see now what aaguid is doing!

File: x/webauthnx/js/webauthn.js
Comment (aeneasr) [inline x/webauthnx/js/webauthn.js:353]: Double submit?

**Review by aeneasr — CHANGES_REQUESTED** [review]
This is in a pretty good

File: x/webauthnx/js/webauthn.js
Comment (hperl) [reply x/webauthnx/js/webauthn.js:137]: This is now only called when the correct node is there. Should the inputs not be there (because of a bug in the UI most likely), the code will simply not execute.

File: ui/node/identifiers.go
Comment (hperl) [reply ui/node/identifiers.go:32]: The corresponding input field holds the browser response to the passkey challenge.

See: https://github.com/ory/kratos/blob/60a1817c6660d1eb668a9994026da385416dcf44/x/webauthnx/js/webauthn.js#L316

File: persistence/sql/migrations/sql/20231108111100000000_credential_types_passkey.up.sql
Comment (hperl) [reply persistence/sql/migrations/sql/20231108111100000000_credential_types_passkey.up.sql:2]: I don't understand how this is important, given that we check the value into git. What is the security consideration here?

File: selfservice/strategy/passkey/passkey_login.go
Comment (hperl) [reply selfservice/strategy/passkey/passkey_login.go:41]: fixed

File: selfservice/strategy/passkey/passkey_registration.go
Comment (hperl) [reply selfservice/strategy/passkey/passkey_registration.go]: Because we need the identifier to properly name the passkey.

File: selfservice/strategy/passkey/passkey_registration.go
Comment (hperl) [reply selfservice/strategy/passkey/passkey_registration.go:111]: Not in a benign case, because the register param is set right before submission.

As for an attacker, what would the gain be?

File: selfservice/strategy/passkey/passkey_registration.go
Comment (hperl) [reply selfservice/strategy/passkey/passkey_registration.go:172]: The display name of the key will be derived from the AAGUID, see here: https://github.com/ory/kratos/blob/60a1817c6660d1eb668a9994026da385416dcf44/identity/credentials_webauthn.go#L36-L39

File: selfservice/strategy/passkey/passkey_registration.go
Comment (hperl) [reply selfservice/strategy/passkey/passkey_registration.go:181]: In this case, `webAuthnSess.UserID` is not user-supplied. It comes from the internal context, which is set here: https://github.com/ory/kratos/blob/60a1817c6660d1eb668a9994026da385416dcf44/selfservice/strategy/passkey/passkey_registration.go#L295-L303

I can add another check to make sure it really is not empty, but it would be an internal server error, as the user can't do anything against it.

File: selfservice/strategy/passkey/passkey_login.go
Comment (hperl) [reply selfservice/strategy/passkey/passkey_login.go:339]: makes sense, fixed :)

File: selfservice/strategy/passkey/passkey_login.go
Comment (hperl) [reply selfservice/strategy/passkey/passkey_login.go:344]: fixed.

File: embedx/config.schema.json
Comment (hperl) [reply embedx/config.schema.json:1642]: Agreed, but the webauthn config also had the legacy single origin, which I did not copy over.

I can also unify this. Your call.

File: identity/extension_credentials.go
Comment (hperl) [reply identity/extension_credentials.go:64]: fixed.

File: ui/node/attributes.go
Comment (hperl) [reply ui/node/attributes.go:106]: fixed

File: ui/node/attributes_input.go
Comment (hperl) [reply ui/node/attributes_input.go:152]: fixed.

File: selfservice/strategy/passkey/passkey_registration.go
Comment (jonas-jonas) [inline selfservice/strategy/passkey/passkey_registration.go:186]: This can panic, no? shouldn't we check that it's actually a non-nil value?

File: selfservice/strategy/passkey/passkey_schema_extension.go
Comment (jonas-jonas) [inline selfservice/strategy/passkey/passkey_schema_extension.go:40]: Why do we ignore the error here?

File: selfservice/strategy/passkey/passkey_strategy.go
Comment (jonas-jonas) [inline selfservice/strategy/passkey/passkey_strategy.go:70]: Why not `strategyDependencies`?

File: test/e2e/cypress/integration/profiles/passwordless/flows.spec.ts
Comment (jonas-jonas) [inline test/e2e/cypress/integration/profiles/passwordless/flows.spec.ts:196]: remove

**Review by jonas-jonas — CHANGES_REQUESTED** [review]
LGTM, just a few minor remarks.

File: test/e2e/cypress/integration/profiles/passwordless/flows.spec.ts
Comment (hperl) [reply test/e2e/cypress/integration/profiles/passwordless/flows.spec.ts:196]: fixed.

File: selfservice/strategy/passkey/passkey_schema_extension.go
Comment (hperl) [reply selfservice/strategy/passkey/passkey_schema_extension.go:40]: Proper validation already happens later. We just want the identifier node, and if it is empty, handle it properly on the call site.

File: selfservice/strategy/passkey/passkey_registration.go
Comment (hperl) [reply selfservice/strategy/passkey/passkey_registration.go:186]: We now check it here: https://github.com/ory/kratos/blob/62c6118bf2e6b3cd984bdda94c084848678e2c6a/selfservice/strategy/passkey/passkey_registration.go#L145

File: selfservice/strategy/passkey/passkey_strategy.go
Comment (hperl) [reply selfservice/strategy/passkey/passkey_strategy.go:70]: This is the standard factory signature, so that we can pass an array of replacement strategies (which we use in Cloud). Otherwise, each factory would have a different signature.

See: https://github.com/ory/kratos/blob/62c6118bf2e6b3cd984bdda94c084848678e2c6a/driver/registry.go#L211-L220

**Review by aeneasr — APPROVED** [review]

**Comment by aeneasr** [comment]
@jonas-jonas are you good with the current state?

**Comment by jonas-jonas** [comment]
@aeneasr I'll re-review now.

File: selfservice/strategy/passkey/passkey_strategy.go
Comment (jonas-jonas) [reply selfservice/strategy/passkey/passkey_strategy.go:70]: ah I see.

**Review by jonas-jonas — APPROVED** [review]
Except for https://github.com/ory/kratos/pull/3656#discussion_r1447420664 this LGTM.

File: selfservice/strategy/passkey/passkey_registration.go
Comment (hperl) [reply selfservice/strategy/passkey/passkey_registration.go:111]: Requiring the method to be set will further complicate submitting the passkey, which is done through a form submit in JS, see: https://github.com/ory/kratos/blob/96c2164e2ad5befa1b05be14b6011556b88ec136/x/webauthnx/js/webauthn.js#L121

We don't require the method to be set for webauthn register requests either, btw: https://github.com/ory/kratos/blob/96c2164e2ad5befa1b05be14b6011556b88ec136/selfservice/strategy/webauthn/registration.go#L94-L114

Comment (hperl) [reply selfservice/strategy/passkey/passkey_registration.go:111]: Specifically, I could add a hidden input with name=method and value=passkey in JS. Is it worth it?

Comment (jonas-jonas) [reply selfservice/strategy/passkey/passkey_registration.go:111]: I'd say it's worth the effort also submitting the method, since we do have the (sometimes loose) convention to submit the method as well. Since this would just be another edge case, potentially causing issues down the road.

But, in the interest of velocity, I'll leave the decision up to you.

**Comment by hperl** [comment]
@jonas-jonas, @aeneasr, two points remain open:

- As I started writing docs, I stumbled over the fact that users need to modify the identity schema to add a `passkey: { identifier: true }` to their schema, most likely to the same identifier as used for webauthn. Do you think there is even a use case for using different identifiers for webauthn and passkey?
- I can't get the coverage up further. Some errors only trigger when the database misbehaves, which would require extensive mocking. How should I proceed? The remaining tests would look more and more artifical.

**Review by jonas-jonas — DISMISSED** [review]

**Comment by hperl** [comment]
I'd like to have one more pair of eyes on the security side of this before merging. @zepatrik maybe?

**Comment by aeneasr** [comment]
I found a few bugs when enabling both webauthn and passkey strategies, probably the triggers get confused. Wrote those things in Slack - would be good to get those fixed in my view :)

**Comment by hperl** [comment]
included in https://github.com/ory/kratos/pull/3748

**Pull request closed without merging**
